extensions/ql-vscode/test/vscode-tests/test-config.ts

```diff
@@ -1,126 +1,225 @@
-      await this.set(
-        this.initialSettingState?.globalValue,
-        ConfigurationTarget.Global,
-      );
-    }
-    if (state?.workspaceValue !== this.initialSettingState?.workspaceValue) {
-      await this.set(
-        this.initialSettingState?.workspaceValue,
-        ConfigurationTarget.Workspace,
-      );
+  private getKey(section: string | undefined, key: string): string {
+    return section ? `${section}.${key}` : key;
+  }
+}
+
+class ChainedInMemoryConfiguration {
+  constructor(private readonly configurations: WorkspaceConfiguration[]) {}
+
+  public getConfiguration(target: ConfigurationTarget) {
+    const configuration = this.configurations.find(
+      (configuration) => configuration.scope === target,
+    );
+
+    if (configuration === undefined) {
+      throw new Error(`Unknown configuration target ${target}`);
     }
-    if (
-      state?.workspaceFolderValue !==
-      this.initialSettingState?.workspaceFolderValue
-    ) {
-      await this.set(
-        this.initialSettingState?.workspaceFolderValue,
-        ConfigurationTarget.WorkspaceFolder,
-      );
+
+    return configuration;
+  }
+
+  public get<T>(section: string | undefined, key: string): T | undefined {
+    for (const configuration of this.configurations) {
+      if (configuration.has(section, key)) {
+        return configuration.get(section, key);
+      }
     }
+    return undefined;
+  }
+
+  public has(section: string | undefined, key: string): boolean {
+    return this.configurations.some((configuration) =>
+      configuration.has(section, key),
+    );
+  }
+
+  public update<T>(
+    section: string | undefined,
+    key: string,
+    value: T,
+    target: ConfigurationTarget,
+  ): void {
+    const configuration = this.getConfiguration(target);
+
+    configuration.update(section, key, value);
   }
 }
 
 // Public configuration keys are the ones defined in the package.json.
 // These keys are documented in the settings page. Other keys are
 // internal and not documented.
-const PKG_CONFIGURATION: Record<string, any> =
-  (function initConfigurationKeys() {
-    // Note we are using synchronous file reads here. This is fine because
-    // we are in tests.
-    const pkg = JSON.parse(
-      readFileSync(join(__dirname, "../../package.json"), "utf-8"),
+const packageConfiguration: Record<
+  string,
+  {
+    default: any | undefined;
+  }
+> = (function initConfigurationKeys() {
+  // Note we are using synchronous file reads here. This is fine because
+  // we are in tests.
+  const pkg = JSON.parse(
+    readFileSync(join(__dirname, "../../package.json"), "utf-8"),
+  );
+  return pkg.contributes.configuration.properties;
+})();
+
+export const vsCodeGetConfiguration = workspace.getConfiguration;
+export let vscodeGetConfigurationMock: jest.SpiedFunction<
+  typeof workspace.getConfiguration
+>;
+
+export const beforeEachAction = async () => {
+  const defaultConfiguration = new DefaultConfiguration(packageConfiguration);
```

Comment:
Would it make more sense to have the default configuration be the _actual_ configuration of the system? I guess the benefit doing it this way is that we are guaranteed that the default config is consistent across all machines and tests.

Reply:
Exactly. I think it's better to make the config fully deterministic rather than also needing to figure out what the actual system configuration is.